sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -527,33 +622,74 @@ class SpillableColumnarBatchHandle private (
     materialized
   }
 
+
+  // used to gate when a spill is actively being done so that a second thread won't
+  // also begin spilling, and a handle won't release the underlying buffer if it's
+  // closed while spilling
+  private var toSpill: Option[ColumnarBatch] = None
+
+  // used to ensure the resource gets cleaned up properly after spilling and avoids
+  // double freeing
+  private var spilled: Option[ColumnarBatch] = None
+  override def releaseDeviceResource(): Unit = {
+    synchronized {
+      super.releaseDeviceResource()
+      spilled.foreach(_.close())
+      spilled = None
+    }
+  }
+
   override def spill(): Long = {
     if (!spillable) {
       0L
     } else {
-      synchronized {
-        if (host.isEmpty && dev.isDefined) {
-          withChunkedPacker { chunkedPacker =>
-            meta = Some(chunkedPacker.getPackedMeta)
-            host = Some(SpillableHostBufferHandle.createHostHandleWithPacker(chunkedPacker))
-          }
-          // We return the size we were created with. This is not the actual size
-          // of this batch when it is packed, and it is used by the calling code
-          // to figure out more or less how much did we free in the device.
-          approxSizeInBytes
+      val thisThreadSpills = synchronized {
+        if (host.isEmpty && dev.isDefined && toSpill.isEmpty) {
+          toSpill = dev
+          GpuColumnVector.incRefCounts(toSpill.get)
+          true
         } else {
-          0L
+          false
         }
       }
+      if (thisThreadSpills) {
+        withChunkedPacker { chunkedPacker =>
+          meta = Some(chunkedPacker.getPackedMeta)
+          var staging: Option[SpillableHostBufferHandle] =
+            Some(SpillableHostBufferHandle.createHostHandleWithPacker(chunkedPacker))
+          synchronized {
+            if (closed) {
+              staging.foreach(_.close())
+              staging = None
+              doClose()
+            } else {
+              host = staging
+            }
+            // set spilled to dev instead of toSpill so that if dev was already closed during spill,
+            // we won't re-close
+            spilled = dev
+            dev = None
+            toSpill = None
+          }
+        }
+        // We return the size we were created with. This is not the actual size
+        // of this batch when it is packed, and it is used by the calling code
+        // to figure out more or less how much did we free in the device.
+        approxSizeInBytes
+      } else {
+        0L
+      }
     }
   }
 
   private def withChunkedPacker[T](body: ChunkedPacker => T): T = {
     val tbl = synchronized {
-      if (dev.isEmpty) {
+      if (toSpill.isEmpty) {
```

Comment:
this is a bad design for this method, and it's my fault. Could we pass to `withChunkedPacker` a batch instead? That way we don't have this exception and don't rely on state.

So the signature would become:
```
private def withChunkedPacker[T](batchToPack: ColumnarBatch)(body: ChunkedPacker => T): T
```